### PR TITLE
Feat: Request Access

### PR DESCRIPTION
**Pull request by DanielHougaard**
# Description 📣

Added ability for users to directly request access to paths / environments defined in a request policy. This is built on top of additional privileges, and it's highly dependent on additional privileges. 

## Type ✨

- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [ ] Documentation

---

- [x] I have read the [contributing guide](https://infisical.com/docs/contributing/getting-started/overview), agreed and acknowledged the [code of conduct](https://infisical.com/docs/contributing/getting-started/code-of-conduct). 📝

### CODE REVIEW

File: backend/src/ee/services/access-approval-request/access-approval-request-service.ts
Comment (dangtony98) [inline backend/src/ee/services/access-approval-request/access-approval-request-service.ts:121]: I don't think this check fully works.

I was able to create multiple of the same access requests with the same contents without hitting the error "You already have a pending access request with the same criteria."

File: backend/src/ee/services/access-approval-request/access-approval-request-service.ts
Comment (dangtony98) [inline backend/src/ee/services/access-approval-request/access-approval-request-service.ts:110]: I don't understand why is this check needed? In what case would `approverUsers.length !== approvers.length"?

File: backend/src/ee/services/access-approval-request/access-approval-request-service.ts
Comment (dangtony98) [inline backend/src/ee/services/access-approval-request/access-approval-request-service.ts:155]: This doesn't have to be in the transaction — We don't need transactions for this function imo.

You can create the access approval request without the transaction. Then, if it fails, it will throw an error anyways which will mean that it won't ever reach `.sendMail()`.

File: backend/src/ee/services/access-approval-request/access-approval-request-service.ts
Comment (dangtony98) [inline backend/src/ee/services/access-approval-request/access-approval-request-service.ts:206]: Nit: This won't scale when the number of requests for a particular project becomes larger; ideally, filtering should be done at an earlier step like with `.where()` in the database call.

OK for the purposes of this PR but keep in mind for future (can deal with the problem when it arises).

File: backend/src/ee/services/access-approval-request/access-approval-request-service.ts
Comment (dangtony98) [inline backend/src/ee/services/access-approval-request/access-approval-request-service.ts:189]: I think we should add authorization resource check to this endpoint function. I don't think everyone should be able to view the list of approval requests for a project.

File: backend/src/ee/services/access-approval-request/access-approval-request-service.ts
Comment (dangtony98) [inline backend/src/ee/services/access-approval-request/access-approval-request-service.ts:255]: This function could've been modified to accept `userIds` instead of project memberships. Doing so would've avoided this additional line since then you could've just passed in `membership.userId`:

```
const reviewerProjectMembership = await projectMembershipDAL.findById(membership.id);
```

File: backend/src/ee/services/access-approval-request/access-approval-request-service.ts
Comment (dangtony98) [inline backend/src/ee/services/access-approval-request/access-approval-request-service.ts:40]: Nit: Incomplete `Pick` for a few data access layers.

File: frontend/src/views/SecretApprovalPage/components/AccessApprovalPolicyList/components/AccessPolicyForm.tsx
Comment (dangtony98) [inline frontend/src/views/SecretApprovalPage/components/AccessApprovalPolicyList/components/AccessPolicyForm.tsx:58]: Nit: `projectSlug` not needed as a param here because you can get it from `useWorkspace()`.

File: frontend/src/views/SecretApprovalPage/components/AccessApprovalRequest/AccessApprovalRequest.tsx
Comment (dangtony98) [inline frontend/src/views/SecretApprovalPage/components/AccessApprovalRequest/AccessApprovalRequest.tsx:1]: Nit: Long file. I would break it down into smaller components/chunks and import them in cuz it's quite difficult to read through otherwise.

For example, the `ReviewRequestModal` can be its own file.

File: frontend/src/views/SecretApprovalPage/components/AccessApprovalRequest/AccessApprovalRequest.tsx
Comment (dangtony98) [inline frontend/src/views/SecretApprovalPage/components/AccessApprovalRequest/AccessApprovalRequest.tsx:267]: Nit: This can definitely be refactored to use `usePopUp`.

Instead of storing data in `selectedRequest`, you can pass in data as second argument into `handlePopUpOpen("reviewRequest");` which you can then access inside the pop-up.

**Review by dangtony98 — COMMENTED** [review]
Mostly nits that can be optionally implemented or disregarded but there are a few changes to be made.

Happy to assist with quick fixes.

File: backend/src/db/schemas/access-approval-requests-reviewers.ts
Comment (DanielHougaard) [reply backend/src/db/schemas/access-approval-requests-reviewers.ts:12]: Taken from secret approval logic.

File: backend/src/ee/services/access-approval-policy/access-approval-policy-service.ts
Comment (DanielHougaard) [reply backend/src/ee/services/access-approval-policy/access-approval-policy-service.ts:140]: Approvers is not a required field.

File: frontend/src/components/basic/Divider.tsx
Comment (DanielHougaard) [reply frontend/src/components/basic/Divider.tsx:1]: This was a quick change, fixed!

File: frontend/src/hooks/api/accessApproval/mutation.tsx
Comment (DanielHougaard) [reply frontend/src/hooks/api/accessApproval/mutation.tsx:41]: I totally agree. This structure was taken from the existing Secret Approval logic. We should probably change this for the secret approvals as well at some point. For now I've gone ahead and changed router grouping to be more REST compliant as suggested!

File: frontend/src/views/SecretApprovalPage/components/AccessApprovalRequest/components/AccessApprovalPolicyRow.tsx
Comment (DanielHougaard) [reply frontend/src/views/SecretApprovalPage/components/AccessApprovalRequest/components/AccessApprovalPolicyRow.tsx:101]: Changed this everywhere we use user.email on the frontend now!

File: backend/src/ee/services/access-approval-policy/access-approval-policy-fns.ts
Comment (DanielHougaard) [reply backend/src/ee/services/access-approval-policy/access-approval-policy-fns.ts:10]: Fixed!

File: backend/src/ee/services/access-approval-policy/access-approval-policy-service.ts
Comment (DanielHougaard) [reply backend/src/ee/services/access-approval-policy/access-approval-policy-service.ts:114]: Fixed.

File: backend/src/ee/services/access-approval-request/access-approval-request-service.ts
Comment (DanielHougaard) [reply backend/src/ee/services/access-approval-request/access-approval-request-service.ts:255]: the `membership` object does not contain a userId, so we'll need to use a dal operation to retrieve it no matter what unfortunately.

File: backend/src/ee/services/access-approval-request/access-approval-request-service.ts
Comment (DanielHougaard) [reply backend/src/ee/services/access-approval-request/access-approval-request-service.ts:121]: Works for me. Let's connect about this tomorrow when we're at the office.

File: backend/src/ee/services/access-approval-request/access-approval-request-service.ts
Comment (DanielHougaard) [reply backend/src/ee/services/access-approval-request/access-approval-request-service.ts:110]: You can consider it a safeguard. The idea behind it was that if an approver deletes their account, it will fail because it will fail to find some of the approver users. But this is redundant because we cascade. If a user is deleted the project membership will be deleted, and if the project membership is deleted, the approver itself is deleted.
_(user -> project membership -> request approver)_

File: backend/src/ee/services/access-approval-request/access-approval-request-service.ts
Comment (DanielHougaard) [reply backend/src/ee/services/access-approval-request/access-approval-request-service.ts:155]: This should be a transaction I'd say. If sendMail fails, something underlying has gone wrong, and the request shouldn't be processed I'd say. It's a safeguard that I wouldn't mind being in place. 

Note: sendMail **won't** fail if it's self-hosted, and the user hasn't set up SMTP.

File: backend/src/ee/services/access-approval-request/access-approval-request-service.ts
Comment (dangtony98) [reply backend/src/ee/services/access-approval-request/access-approval-request-service.ts:127]: Found the edge-case to replicate:
- Suppose User A is requesting access and User B can approve.
- User A requests access to any environment/path.
- User B rejects the request from User A.
- User A can now create multiple of the same access requests to the environment/path.

Comment (DanielHougaard) [reply backend/src/ee/services/access-approval-request/access-approval-request-service.ts]: Great catch! This is now fixed.

**Review by dangtony98 — APPROVED** [review]

**Comment by dangtony98** [comment]
@DanielHougaard We'll merge this when you're back together with Maidul.

**Comment by vmatsiiako** [comment]
When are we thinking to merge this?

**Comment by maidul98** [comment]
Let's get this done as soon as possible next week otheriwse it will get too stale

**Review by maidul98 — APPROVED** [review]
note: merging without group support. Group support will come at a later time

**Pull request merged**
